### PR TITLE
Extract classes from Resolver.cs and NameResolutionAndTypeInference.cs

### PR DESCRIPTION
**Pull request by keyboardDrummer**
<small>By submitting this pull request, I confirm that my contribution is made under the terms of the [MIT license](https://github.com/dafny-lang/dafny/blob/master/LICENSE.txt).</small>

### CODE REVIEW

**Comment by alex-chew** [comment]
LGTM up to and including commit e5c3b80, and I will come back to review the remaining few commits.

File: Source/DafnyCore/Resolver/NameResolutionAndTypeInference/NameResolutionAndTypeInference.cs [hunk 1]
```diff
@@ -5852,6 +5852,8 @@ public record ResolveTypeReturn(Type ReplacementType, ExprDotName LastComponent)
       return rr;
     }
 
+    public record MethodCallInformation(IToken Tok, MemberSelectExpr Callee, List<ActualBinding> ActualParameters);
```
Comment (alex-chew) [inline Source/DafnyCore/Resolver/NameResolutionAndTypeInference/NameResolutionAndTypeInference.cs:5855]: Was the `ObjectInvariant` method not being used, or should we add it to this?

**Review by alex-chew — COMMENTED** [review]
Mostly LGTM. The indentation changes in XConstraint seem to greatly confuse the diff algorithm, and by extension, me...

File: Source/DafnyCore/Resolver/NameResolutionAndTypeInference/NameResolutionAndTypeInference.cs (hunk 1 above)
Comment (keyboardDrummer) [reply Source/DafnyCore/Resolver/NameResolutionAndTypeInference/NameResolutionAndTypeInference.cs:5855]: `[ContractInvariantMethod]` is a Noop, so indeed `ObjectInvariant()` was purely for documentation purposes. I don't think we should keep it though.

**Comment by keyboardDrummer** [comment]
> The indentation changes in XConstraint seem to greatly confuse the diff algorithm, and by extension, me...

I don't recall making any changes in XConstraint, besides extracting it to a separate file which changed the indentation.

File: Source/DafnyCore/Resolver/Resolver.cs [hunk 2]
```diff
@@ -5818,6 +4612,7 @@ public class ResolveTypeOption {
     /// context. 
     /// </summary>
     public void CheckLocalityUpdates(Statement stmt, ISet<LocalVariable> localsAllowedInUpdates, string where) {
+      // TODO it looks like this method has no side-effects and doesn't return anything.
```
Comment (atomb) [inline Source/DafnyCore/Resolver/Resolver.cs:4615]: Mysterious. We should probably investigate that more deeply at some point, so I'm adding a comment to make it slightly more visible. Given that this is a purely refactoring PR, I don't think we need to include resolving it here, though.

File: Source/DafnyCore/Resolver/NameResolutionAndTypeInference/NameResolutionAndTypeInference.cs (hunk 1 above)
Comment (atomb) [reply Source/DafnyCore/Resolver/NameResolutionAndTypeInference/NameResolutionAndTypeInference.cs:5869]: Yeah, it'd be nice to start taking more advantage of those features in this code base.

**Review by atomb — APPROVED** [review]
I haven't compared the source and destination text of the classes you extracted, but assuming those are all pure extractions I think this looks good.

**Pull request merged**